docs/block-components.md

```diff
@@ -0,0 +1,345 @@
+					onFocus: props.setFocus
+				}
+			)
+		];
+	},
+
+	save: function( props ) {
+		var content = props.attributes.content;
+
+		return el( 'p', { className: props.className }, content );
+	},
+} );
+```
+{% ESNext %}
+```js
+const {
+	registerBlockType,
+	Editable,
+	BlockControls,
+	AlignmentToolbar,
+	source
+} = wp.blocks;
+const { children } = source;
+
+registerBlockType( 'gutenberg-boilerplate-esnext/hello-world-step-04', {
+	title: 'Hello World (Step 4)',
+
+	icon: 'universal-access-alt',
+
+	category: 'layout',
+
+	attributes: {
+		content: {
+			type: 'array',
+			source: children( 'p' ),
+		},
+	},
+
+	edit( { attributes, setAttributes, focus } ) {
+		const { content, alignment } = attributes;
+
+		function onChangeContent( newContent ) {
+			setAttributes( { content: newContent } );
+		}
+
+		function onChangeAlignment( newAlignment ) {
+			setAttributes( { alignment: newAlignment } );
+		}
+
+		return [
+			!! focus && (
+				<BlockControls key="controls">
+					<AlignmentToolbar
+						value={ alignment }
+						onChange={ onChangeAlignment }
+					/>
+				</BlockControls>
+			),
+			<Editable
+				key="editable"
+				tagName="p"
+				className={ props.className }
+				style={ { textAlign: alignment } }
+				onChange={ onChangeContent }
+				value={ content }
+				focus={ focus }
+				onFocus={ props.setFocus }
+			/>
+		];
+	},
+
+	save( { attributes, className } ) {
+		const { content } = attributes;
+
+		return <p className={ className }>{ content }</p>;
+	},
+} );
+```
+{% end %}
+
+Note that you should only include `BlockControls` if the block is currently selected. We must test that the `focus` value is truthy before rendering the element, otherwise you will inadvertently cause controls to be shown for the incorrect block type.
+
+### BlockAlignmentToolbar
+
+...
+
+## Inspector
+
+While the toolbar area is useful for displaying controls to toggle attributes of a block, sometimes a block needs to include extra functionality and settings that are not appropriate to show on the block itself or that require more screen space. Gutenberg includes a block inspector region in the sidebar for this purpose.
+
+### InspectorControls
+
+Similar to rendering a toolbar, if you include an `InspectorControls` element in the return value of your block type's `edit` function, those controls will be shown in the inspector region. All the children passed to this component will appear in the sidebar when the tab and the corresponding block are selected.
+
+### BlockDescription
+
+A block can display a short text describing what its purpose is. It is recommended to always include one when using the inspector.
+
+# Components
+
+In order to make it easier to create sophisticated blocks quickly, reducing the amount of work you have to do to develop common functionality, several components have been created which can be accessed through the `wp.components` global. This section covers the available UI components.
+
+## Placeholder
```

Comment:
Just an idea, should we maybe keep the component related docs in their own README files and only cross-link them in here with a very brief description?

Reply:
I'd prefer to have them here as this will easier to navigate and refer to.

Reply:
A middle-ground could be to create a separate page for each component using its README